elements/lisk-p2p/src/peer_selection.ts

```diff
@@ -104,5 +104,38 @@ export const selectPeersForConnection = (
 		return [];
 	}
 
-	return shuffle(input.peers).slice(0, input.peerLimit);
+	if (
+		input.peerLimit === undefined ||
+		input.peerLimit >= input.triedPeers.length + input.newPeers.length
+	) {
+		return [...input.newPeers, ...input.triedPeers];
+	}
+
+	if (input.triedPeers.length === 0 && input.newPeers.length === 0) {
+		return [];
+	}
+
+	const x =
+		input.triedPeers.length / (input.triedPeers.length + input.newPeers.length);
+	// tslint:disable-next-line: no-magic-numbers
+	const r = Math.max(x, 0.5);
+
+	const shuffledTriedPeers = shuffle(input.triedPeers);
+	const shuffledNewPeers = shuffle(input.newPeers);
+
+	return [...Array(input.peerLimit)].map(() => {
+		if (shuffledTriedPeers.length !== 0) {
+			if (Math.random() < r) {
+				// With probability r
+				return shuffledTriedPeers.splice(0, 1)[0];
```

Comment:
`Array.prototype.splice` is an expensive operation because each time you call splice, it has to update the indices of all remaining elements in the array. It's expensive to do this for every peer.

Maybe we can use the current index from the `map` operation instead?